### PR TITLE
ansible: install temurin JDK on macOS

### PR DESCRIPTION
**Pull request by targos**
Refs: https://github.com/nodejs/build/issues/3030

### CODE REVIEW

**Comment by targos** [comment]
I have never used Ansible before and I don't to break everything.
What's the safest way to test this?

Also, since this is completely changing how the package is installed, are we supposed to uninstall the previous stuff? How is that done?

**Comment by richardlau** [comment]
> I have never used Ansible before and I don't to break everything. What's the safest way to test this?

Ansible has a [`--check` mode](https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html) which will attempt to run the playbook without making changes. The caveat here is whether we're using raw commands in our playbook and then depending on the output of those commands -- Ansible doesn't know whether those commands have side effects so doesn't run them in check mode and then the things that depend on them subsequently fail. I've fixed up some of these in the past (https://github.com/nodejs/build/pull/2682) but it's possible more have crept in and would need to be fixed up.

Also [`ansible-playbook` has ` --limit` (or `-l`) option](https://docs.ansible.com/ansible/latest/user_guide/intro_patterns.html#patterns-and-ansible-playbook-flags) to restrict which host the playbook is running against. So, for example, 
```
ansible-playbook --limit "test-nearform-macos10.15-x64-1" ansible/playbooks/jenkins/worker/create.yml
```
would only run the playbook against `test-nearform-macos10.15-x64-1`. Limiting to one host first will contain the changes in case something goes awry -- once you're confident the changes are correct you can then rerun with a wider glob for `--limit` (or do the other machines individually).

> Also, since this is completely changing how the package is installed, are we supposed to uninstall the previous stuff? How is that done?

You can run one-off individual Ansible tasks with `ansible`, 
e.g. here's one of the commands I was using for https://github.com/nodejs/build/pull/3062 to uninstall the python chocolatey package on a subset of the Windows machines prior to running the playbook:
```
ansible -m win_chocolatey -a "name=python state=absent remove_dependencies=yes" -v test-rackspace-win2012r2_vs2015-x64-?
```
You can also just ssh into the machine directly and remove the old package -- the idea of the Ansible scripts is that we can run them against a new/fresh image (which presumably will not have the old package) and the script should put the image into a usable state.

**Comment by targos** [comment]
I took https://ci.nodejs.org/computer/test-nearform-macos11.0-arm64-1/ offline to update it.

**Comment by targos** [comment]
Failed:

```
TASK [java-base : Check if java is already installed] ******************************************************************
[WARNING]: raw module does not support the environment keyword
fatal: [test-nearform-macos11.0-arm64-1]: FAILED! => {"changed": false, "msg": "non-zero return code", "rc": 1, "stderr": "Shared connection to 83.147.191.75 closed.\r\n", "stderr_lines": ["Shared connection to 83.147.191.75 closed."], "stdout": "The operation couldn’t be completed. Unable to locate a Java Runtime.\r\nPlease visit http://www.java.com for information on installing Java.\r\n\r\n", "stdout_lines": ["The operation couldn’t be completed. Unable to locate a Java Runtime.", "Please visit http://www.java.com for information on installing Java.", ""]}
...ignoring

TASK [java-base : install java] ****************************************************************************************
skipping: [test-nearform-macos11.0-arm64-1]

TASK [java-base : install java (macOS)] ********************************************************************************
fatal: [test-nearform-macos11.0-arm64-1]: FAILED! => {"msg": "Failed to set permissions on the temporary files Ansible needs to create when becoming an unprivileged user (rc: 1, err: chmod: Invalid file mode: A+user:administrator:rx:allow\n}). For information on working around this, see https://docs.ansible.com/ansible-core/2.11/user_guide/become.html#becoming-an-unprivileged-user"}
```

**Comment by targos** [comment]
Oh I see, the problem is that on this mac, the administrator account is called `admin`, not `administrator`. Is it possible to `become_user` with a variable somehow?

**Comment by targos** [comment]
Can I use `become_user: {{ ansible_user }}`? I'm wondering why it was hardcoded in this place.

**Comment by AshCripps** [comment]
you are going to hit a caveat on some machines where the user is sometimes `admin` and other times `administrator` IIRC any nearform machine we didnt setup is `admin` and any machine I created is `administrator`. 

My work around was to just change the user to admin run the ansible and then change it back.

**Comment by targos** [comment]
Install looks good!

```
admin@test-nearform-macos11 ~ % java -version
openjdk version "19.0.1" 2022-10-18
OpenJDK Runtime Environment Temurin-19.0.1+10 (build 19.0.1+10)
OpenJDK 64-Bit Server VM Temurin-19.0.1+10 (build 19.0.1+10, mixed mode)
```

**Comment by targos** [comment]
I also updated `test-nearform-macos10.15-x64-1` and `test-nearform-macos10.15-x64-2`.
Now before we go further, are we OK with installing the latest Java or do we want to stay on LTS? For that, I'd have to tap `homebrew/cask-versions` and install `temurin17` instead of `temurin`.

**Comment by richardlau** [comment]
> Can I use `become_user: "{{ ansible_user }}"`? I'm wondering why it was hardcoded in this place.

Yes, that should work. We use it elsewhere, e.g. https://github.com/nodejs/build/blob/b02d18741a07418eaeb53763b0ac6b5df9d7f936/ansible/roles/package-upgrade/tasks/partials/brew.yml#L28-L29

**Comment by sxa** [comment]
> > Now before we go further, are we OK with installing the latest Java or do we want to stay on LTS?
> @sxa thoughts?

I'd stick with LTS (17) personally for this.

**Comment by targos** [comment]
I have an unrelated problem with `test-orka-macos11-x64-2`:

```
failed: [test-orka-macos11-x64-2] (item=cmake,python,ccache) => {"ansible_loop_var": "package", "changed": false, "msg": "Warning: python@3.9 3.9.13_1 already installed.", "package": "cmake,python,ccache"}
```

**Comment by targos** [comment]
# Update status

## Updated to Java 17

- release-nearform-macos10.15-x64-1
- release-nearform-macos11.0-arm64-1
- release-orka-macos10.15-x64-1
- test-nearform-macos10.15-x64-1
- test-nearform-macos10.15-x64-2
- test-nearform-macos11.0-arm64-1
- test-orka-macos10.14-x64-1
- test-orka-macos10.14-x64-2
- test-orka-macos10.14-x64-3
- test-orka-macos11-x64-1
- test-orka-macos11-x64-2

## Impossible to connect via SSH

- release-macstadium-macos11.0-arm64-1
- test-macstadium-macos11.0-arm64-3
- test-macstadium-macos11.0-arm64-4

## Offline

- test-nearform-macos10.15-x64-3
- test-orka-macos10.15-x64-1
- test-orka-macos10.15-x64-2

**Comment by targos** [comment]
@AshCripps What am I missing to connect to the macstadium machines?
This is what I get when I try to ssh:

```
$ ssh test-macstadium-macos11.0-arm64-3
Connection closed by 207.254.38.86 port 22
```

**Comment by AshCripps** [comment]
@targos I think theyve been down for a while - Ill raise a ticket with macstadium to investigate.

**Comment by AshCripps** [comment]
SERVICE-146218 raised. It seems they have updated their portal and we can do things like hard reboots etc bu tthat requires a setting to be turned that I dont know if we have so Ive just raised a ticket for now.

File: ansible/roles/package-upgrade/tasks/partials/brew.yml [hunk 1]
```diff
@@ -44,11 +45,11 @@
     become_user: "{{ ansible_user }}"
     homebrew_tap:
       name: buo/cask-upgrade
-
-  - name: Add AdoptOpenJDK Java Repo
+
```
Comment (targos) [inline ansible/roles/package-upgrade/tasks/partials/brew.yml:48]:
```suggestion

```

**Review by richardlau — APPROVED** [review]

**Pull request merged**